src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,115 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double COEFFICIENT = 1.5;
+
+    private T[] array = (T[]) new Object[INITIAL_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexBounds(index < 0);
+        ensureCapacity();
+        if (index == size) {
+            add(value);
+        } else {
+            T t = array[index];
+            checkIndexBounds(t == null);
+            System.arraycopy(array, index, array, index + 1, size - index);
+            array[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexBounds(isCorrectIndex(index));
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexBounds(isCorrectIndex(index));
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexBounds(isCorrectIndex(index));
+        T t = array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        size--;
+        return t;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (element == null && array[i] == null) {
+                size--;
+                return null;
+            }
+            if (array[i] != null && array[i].equals(element)) {
+                index = i;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException();
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isFull() {
+        return size == array.length - 1;
+    }
+
+    private T[] grow() {
+        T[] newArray = (T[]) new Object[(int) (array.length * COEFFICIENT)];
+        System.arraycopy(array, 0, newArray, 0, array.length);
+        return newArray;
```

Comment:
The grow method should be responsible for resizing the array and should not return any value. The current implementation returns the new array but does not actually update the 'array' field. This is misleading and does not follow the Single Responsibility Principle.